pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,869 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package capacityscheduling
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	policy "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	policylisters "k8s.io/client-go/listers/policy/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+	extenderv1 "k8s.io/kube-scheduler/extender/v1"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/scheduler/core"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/defaultpreemption"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+	"k8s.io/kubernetes/pkg/scheduler/util"
+	cutil "sigs.k8s.io/scheduler-plugins/pkg/capacityscheduling/util"
+
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/config"
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+	"sigs.k8s.io/scheduler-plugins/pkg/generated/clientset/versioned"
+	"sigs.k8s.io/scheduler-plugins/pkg/generated/informers/externalversions"
+)
+
+// CapacityScheduling is a plugin that implements the mechanism of capacity scheduling.
+type CapacityScheduling struct {
+	frameworkHandle framework.FrameworkHandle
+	pdbLister       policylisters.PodDisruptionBudgetLister
+	elasticQuotas   map[string]*ElasticQuotaInfo
+	assumedPods     map[string]bool
+	lock            *sync.RWMutex
+}
+
+// PreFilterState computed at PreFilter and used at PostFilter or Reserve.
+type PreFilterState struct {
+	framework.Resource
+}
+
+// Clone the prefilter state.
+func (s *PreFilterState) Clone() framework.StateData {
+	return s
+}
+
+// SnapshotState stores the snapshot of elasticQuotas.
+type SnapshotState struct {
+	elasticQuotas map[string]ElasticQuotaInfo
+}
+
+// Clone the Snapshot state.
+func (s *SnapshotState) Clone() framework.StateData {
+	return s
+}
+
+var _ framework.PreFilterPlugin = &CapacityScheduling{}
+var _ framework.PostFilterPlugin = &CapacityScheduling{}
+var _ framework.ReservePlugin = &CapacityScheduling{}
+
+const (
+	// Name is the name of the plugin used in Registry and configurations.
+	Name = "CapacityScheduling"
+	// preFilterStateKey is the key in CycleState to NodeResourcesFit pre-computed data.
+	// Using the name of the plugin will likely help us avoid collisions with other plugins.
+	preFilterStateKey            = "PreFilter" + Name
+	SnapshotStateKey             = "SnapshotState"
+	defaultKubeConfigPath string = "/etc/kubernetes/scheduler.conf"
+)
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (c *CapacityScheduling) Name() string {
+	return Name
+}
+
+type PluginConfig struct {
+	kubeconfig string
+}
```

Comment:
It's already defined in apis/config/types.go. This one should be deleted.

Reply:
@denkensk?

Reply:
Done